### Import Necessary Libaries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner

!conda install -c conda-forge geopy --yes # Uncomment out if you need to install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values


# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes # Uncomment out you need to install folium
import folium # plotting library

!pip install xlrd


print('Folium installed')
print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.11

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          90 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.20.0-py_0 conda-forge


geopy-1.20.0         | 57 KB     | ##################################### | 100% 
geographiclib-1.49   | 

After importing/installing all the necessary packages, we will read the excel spreadsheet. The neighborhood and location columns are from wikipedia. I added the city, state, and combo column to allow geolocator to work in the following code.

In [2]:
# Reading table into dataframe
bmore_df = pd.read_excel('Baltimore_Neigh_List.xlsx')
print(bmore_df.shape)
bmore_df

(221, 5)


Neighborhood       City   Location State  \
0                Abell  Baltimore      North    MD   
1            Allendale  Baltimore       West    MD   
2              Arcadia  Baltimore  Northeast    MD   
3            Arlington  Baltimore  Northwest    MD   
4    Armistead Gardens  Baltimore  Northeast    MD   
..                 ...        ...        ...   ...   
216           Woodring  Baltimore  Northeast    MD   
217         Wyman Park  Baltimore      North    MD   
218          Wyndhurst  Baltimore      North    MD   
219       Yale Heights  Baltimore  Southwest    MD   
220          York Road  Baltimore  Northeast    MD   

                                Combo  
0                Abell, Baltimore, MD  
1            Allendale, Baltimore, MD  
2              Arcadia, Baltimore, MD  
3            Arlington, Baltimore, MD  
4    Armistead Gardens, Baltimore, MD  
..                                ...  
216           Woodring, Baltimore, MD  
217         Wyman Park, Baltimore, MD  
218          Wyndhurst, Baltimore, MD  
219       Yale Heights, Baltimore, MD  
220          York Road, Baltimore, MD  

[221 rows x 5 columns]

Here we are reading the data in the combo column using a for-loop

In [3]:
# allows to read the specific column of the spreadsheet
def readcolumn(filename,column):
    #select sheet name and selct column as index,index_col=0
    df = pd.read_excel(filename)
    headername = list(df)
    print(headername)
    column_data =df[list(df)[column]].tolist()
    return  column_data

neigh_list = readcolumn('Baltimore_Neigh_List.xlsx', 4)

['Neighborhood', 'City', 'Location', 'State', 'Combo']


Since the city names are not unique, I had to add in 'Baltimore' to the code so the geolocator would pull the correct lat/long coordinates.

In [13]:
# Prints out the neighborhood list with the latitude and longitude
df = []
for neigh in neigh_list:
    try:
        geolocator = Nominatim(user_agent="bmore_explorer")
        location = geolocator.geocode(neigh)
        latitude = location.latitude
        longitude = location.longitude
        print(neigh, latitude, longitude)
        df.append({'Location':location,
                  'Latitude': latitude,
                  'Longitude': longitude})
    except Exception as e:
        print('Error, skipping address...', e)

Abell, Baltimore, MD 39.3261388 -76.611737
Allendale, Baltimore, MD 39.2916866 -76.6769424
Arcadia, Baltimore, MD 39.3340564 -76.5664996
Arlington, Baltimore, MD 39.3486919 -76.6826661
Armistead Gardens, Baltimore, MD 39.3066459 -76.5562793455641
Arundel Cove, Baltimore, MD 39.1985156 -76.5673626074159
Ashburton, Baltimore, MD 39.3279621 -76.6710811
Baltimore Highlands, Baltimore, MD 39.2331629 -76.636634
Barclay, Baltimore, MD 39.3149545 -76.6122023
Barre Circle, Baltimore, MD 39.2848699 -76.6276523
Bayview, Baltimore, MD 39.2894325 -76.5416232
Beechfield, Baltimore, MD 39.2752308 -76.6975606
Belair-Edison, Baltimore, MD 39.3215607 -76.5678685
Belair-Parkside, Baltimore, MD 39.327438 -76.566113
Berea, Baltimore, MD 39.3092351 -76.5779746
Beverly Hills, Baltimore, MD 39.3402861 -76.5694739
Biddle Street, Baltimore, MD 39.3093234 -76.5260695
Bolton Hill, Baltimore, MD 39.3084715 -76.6259787
Brewer's Hill, Baltimore, MD 39.2838792 -76.5645376
Broadway East, Baltimore, MD 39.308818 -76.58

In [14]:
# Convert list to dataframe
df = pd.DataFrame.from_dict (df)
df

Location   Latitude  Longitude
0    (Abell, Baltimore, Maryland, 21218, USA, (39.3...  39.326139 -76.611737
1    (Allendale, Baltimore, Maryland, 21229, USA, (...  39.291687 -76.676942
2    (Arcadia, Baltimore, Maryland, 21082:21234, US...  39.334056 -76.566500
3    (Arlington, Baltimore, Maryland, 21215, USA, (...  39.348692 -76.682666
4    (Armistead Gardens, Baltimore, Maryland, USA, ...  39.306646 -76.556279
..                                                 ...        ...        ...
216  (Woodring, Baltimore, Maryland, 21256, USA, (3...  39.363819 -76.544286
217  (Wyman Park, Baltimore, Maryland, 21211, USA, ...  39.333540 -76.626371
218  (Wyndhurst, Baltimore, Maryland, 21210, USA, (...  39.355871 -76.631357
219  (Yale Heights, Baltimore, Maryland, 21229, USA...  39.275706 -76.689895
220  (York Road, Kernewood, Baltimore, Maryland, 21...  39.346172 -76.609618

[221 rows x 3 columns]

Let's merge the dataframes!

In [15]:
# Merging the two dataframes
df_merged = df.merge(bmore_df, how='outer', left_index=True, right_index=True)
df_merged

Location_x   Latitude  Longitude  \
0    (Abell, Baltimore, Maryland, 21218, USA, (39.3...  39.326139 -76.611737   
1    (Allendale, Baltimore, Maryland, 21229, USA, (...  39.291687 -76.676942   
2    (Arcadia, Baltimore, Maryland, 21082:21234, US...  39.334056 -76.566500   
3    (Arlington, Baltimore, Maryland, 21215, USA, (...  39.348692 -76.682666   
4    (Armistead Gardens, Baltimore, Maryland, USA, ...  39.306646 -76.556279   
..                                                 ...        ...        ...   
216  (Woodring, Baltimore, Maryland, 21256, USA, (3...  39.363819 -76.544286   
217  (Wyman Park, Baltimore, Maryland, 21211, USA, ...  39.333540 -76.626371   
218  (Wyndhurst, Baltimore, Maryland, 21210, USA, (...  39.355871 -76.631357   
219  (Yale Heights, Baltimore, Maryland, 21229, USA...  39.275706 -76.689895   
220  (York Road, Kernewood, Baltimore, Maryland, 21...  39.346172 -76.609618   

          Neighborhood       City Location_y State  \
0                Abell  Baltimore      North    MD   
1            Allendale  Baltimore       West    MD   
2              Arcadia  Baltimore  Northeast    MD   
3            Arlington  Baltimore  Northwest    MD   
4    Armistead Gardens  Baltimore  Northeast    MD   
..                 ...        ...        ...   ...   
216           Woodring  Baltimore  Northeast    MD   
217         Wyman Park  Baltimore      North    MD   
218          Wyndhurst  Baltimore      North    MD   
219       Yale Heights  Baltimore  Southwest    MD   
220          York Road  Baltimore  Northeast    MD   

                                Combo  
0                Abell, Baltimore, MD  
1            Allendale, Baltimore, MD  
2              Arcadia, Baltimore, MD  
3            Arlington, Baltimore, MD  
4    Armistead Gardens, Baltimore, MD  
..                                ...  
216           Woodring, Baltimore, MD  
217         Wyman Park, Baltimore, MD  
218          Wyndhurst, Baltimore, MD  
219       Yale Heights, Baltimore, MD  
220          York Road, Baltimore, MD  

[221 rows x 8 columns]

In [16]:
# Removing the whitespace and renaming
df_merged.columns = ['location', 'Latitude', 'Longitude', 'Neighborhood', 'City', 'Location', 'State', 'Combo', ]
print(df_merged.columns.tolist())

['location', 'Latitude', 'Longitude', 'Neighborhood', 'City', 'Location', 'State', 'Combo']


I tried to drop the columns I did not need, but was running into an error, so I just created a new dataframe to bypass the error.

In [17]:
# Creating a new dataframe with the necessary columns
df_final = df_merged[['Neighborhood', 'Location', 'Latitude', 'Longitude']]
df_final

Neighborhood   Location   Latitude  Longitude
0                Abell      North  39.326139 -76.611737
1            Allendale       West  39.291687 -76.676942
2              Arcadia  Northeast  39.334056 -76.566500
3            Arlington  Northwest  39.348692 -76.682666
4    Armistead Gardens  Northeast  39.306646 -76.556279
..                 ...        ...        ...        ...
216           Woodring  Northeast  39.363819 -76.544286
217         Wyman Park      North  39.333540 -76.626371
218          Wyndhurst      North  39.355871 -76.631357
219       Yale Heights  Southwest  39.275706 -76.689895
220          York Road  Northeast  39.346172 -76.609618

[221 rows x 4 columns]

In [18]:
# Ensure the types are correct for mapping
print(df_final.dtypes)

Neighborhood     object
Location         object
Latitude        float64
Longitude       float64
dtype: object


We will now see if any data was not merged correctly or if any is not present

In [19]:
# Calculate how many rows are missing
missing_data = df_final.isnull()

for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")  

Neighborhood
False    221
Name: Neighborhood, dtype: int64

Location
False    221
Name: Location, dtype: int64

Latitude
False    221
Name: Latitude, dtype: int64

Longitude
False    221
Name: Longitude, dtype: int64



It looks like all of the data was transferred correctly. 221 out of 221 rows!

# -------------------------------------------------------------------------------------

## Explore and cluster the neighborhoods in Baltimore

In [20]:
# create map of Baltimore using latitude and longitude values from above
map_baltimore = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map from 'can_loc' df
for lat, lng, location, neighborhood in zip(df_final['Latitude'], df_final['Longitude'], df_final['Location'], df_final['Neighborhood']):
    label = '{}, {}'.format(location, neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_baltimore)  
    
map_baltimore

In [21]:
# View Neighborhood's total by cardinal direction
pd.value_counts(df_final['Location'].values, sort=False)

Northeast    31
Southeast    23
West         18
Southwest    26
Central      18
Northwest    27
North        44
East         16
South        18
dtype: int64

### Define credentials

In [22]:
CLIENT_ID = 'ENTER YOUR CLIENT_ID HERE' # your Foursquare ID
CLIENT_SECRET = 'ENTER YOUR CLIENT_SECRET HERE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + 'ENTERED')
print('CLIENT_SECRET:' + 'ENTERED')

Your credentails:
CLIENT_ID: ENTERED
CLIENT_SECRET:ENTERED


Finding the top 100 venues around Baltimore

In [23]:
radius = 500
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
print('URL entered with credentials')

URL entered with credentials


In [24]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d7132f666fc65002c928719'},
 'response': {'venues': [{'id': '4e5235e9b61c097de677359c',
    'name': 'MTA #48  York and Coldspring',
    'location': {'lat': 39.34611835982064,
     'lng': -76.60937547761814,
     'labeledLatLngs': [{'label': 'display',
       'lat': 39.34611835982064,
       'lng': -76.60937547761814}],
     'distance': 21,
     'cc': 'US',
     'state': 'Maryland',
     'country': 'United States',
     'formattedAddress': ['Maryland', 'United States']},
    'categories': [{'id': '52f2ab2ebcbc57f1066b8b4f',
      'name': 'Bus Stop',
      'pluralName': 'Bus Stops',
      'shortName': 'Bus Stop',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/busstation_',
       'suffix': '.png'},
      'primary': True}],
    'referralId': 'v-1567699702',
    'hasPerk': False},
   {'id': '4e7cc57b483b1e7b5ae348ef',
    'name': 'Chicken Run',
    'location': {'address': '4711 York Rd',
     'lat': 39.346191162312884,
     'lng

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Location', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Since we want to find the popular venues by location, we are going to first analyze the data through geographical location and will later analyze each neighborhood.

In [26]:
baltimore_venues = getNearbyVenues(names=df_final['Location'],
                                   latitudes=df_final['Latitude'],
                                   longitudes=df_final['Longitude']
                                  )

North
West
Northeast
Northwest
Northeast
South
Northwest
Southeast
East
Southwest
Southeast
Southwest
Northeast
Northeast
East
Northeast
East
Central
Southeast
East
South
South
Southeast
Northwest
North
Southeast
Southwest
Central
North
Northeast
Northeast
Northwest
Central
North
North
South
Northwest
Northeast
North
West
Northwest
South
North
East
Southwest
Northwest
Northwest
Central
East
Northwest
East
West
Southeast
Southwest
West
Northeast
East
North
Southwest
Northwest
South
Southeast
Southeast
Northwest
Northeast
West
Southwest
Northeast
East
Northwest
North
Northeast
North
Southeast
Southeast
East
North
Northwest
North
Northwest
North
Northwest
Southeast
Southeast
South
West
North
South
West
Southeast
Northeast
Southwest
North
Northeast
Southwest
North
Northwest
Northeast
Central
Southwest
East
Southeast
Southeast
North
North
North
Southeast
North
North
South
Northwest
Northeast
North
West
Central
Northwest
Southeast
South
North
Northwest
Central
Central
South
Northeast
East
No

### Examine the new dataframe

Now lets view the new dataframe and shape

In [27]:
print(baltimore_venues.shape)
baltimore_venues.head()

(3879, 7)


Location  Neighborhood Latitude  Neighborhood Longitude  \
0    North              39.326139              -76.611737   
1    North              39.326139              -76.611737   
2    North              39.326139              -76.611737   
3    North              39.326139              -76.611737   
4    North              39.326139              -76.611737   

                         Venue  Venue Latitude  Venue Longitude  \
0  32nd Street Farmer's Market       39.327362       -76.610851   
1                Pete's Grille       39.327064       -76.609593   
2               The Book Thing       39.325253       -76.610272   
3     Normal's Books & Records       39.326012       -76.609903   
4      Peabody Heights Brewery       39.324744       -76.610819   

   Venue Category  
0  Farmers Market  
1  Breakfast Spot  
2       Bookstore  
3     Record Shop  
4         Brewery

In [28]:
# one hot encoding
baltimore_onehot = pd.get_dummies(baltimore_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
baltimore_onehot['Location'] = baltimore_venues['Location']

In [29]:
# move neighborhood column to the first column
neighborhood = baltimore_onehot['Location']
baltimore_onehot.drop(labels=['Location'], axis=1,inplace = True)
baltimore_onehot.insert(0, 'Location', neighborhood)
baltimore_onehot

Location  ATM  Adult Boutique  Afghan Restaurant  African Restaurant  \
0         North    0               0                  0                   0   
1         North    0               0                  0                   0   
2         North    0               0                  0                   0   
3         North    0               0                  0                   0   
4         North    0               0                  0                   0   
...         ...  ...             ...                ...                 ...   
3874  Northeast    0               0                  0                   0   
3875  Northeast    0               0                  0                   0   
3876  Northeast    0               0                  0                   0   
3877  Northeast    0               0                  0                   0   
3878  Northeast    0               0                  0                   0   

      Airport  American Restaurant  Antique Shop  Aquarium  \
0           0                    0             0         0   
1           0                    0             0         0   
2           0                    0             0         0   
3           0                    0             0         0   
4           0                    0             0         0   
...       ...                  ...           ...       ...   
3874        0                    0             0         0   
3875        0                    0             0         0   
3876        0                    0             0         0   
3877        0                    0             0         0   
3878        0                    0             0         0   

      Argentinian Restaurant  ...  Vietnamese Restaurant  Volleyball Court  \
0                          0  ...                      0                 0   
1                          0  ...                      0                 0   
2                          0  ...                      0                 0   
3                          0  ...                      0                 0   
4                          0  ...                      0                 0   
...                      ...  ...                    ...               ...   
3874                       0  ...                      0                 0   
3875                       0  ...                      0                 0   
3876                       0  ...                      0                 0   
3877                       0  ...                      0                 0   
3878                       0  ...                      0                 0   

      Waterfall  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0             0                   0           0         0          0   
1             0                   0           0         0          0   
2             0                   0           0         0          0   
3             0                   0           0         0          0   
4             0                   0           0         0          0   
...         ...                 ...         ...       ...        ...   
3874          0                   0           0         0          0   
3875          0                   0           0         0          0   
3876          0                   0           0         0          0   
3877          0                   0           0         0          0   
3878          0                   0           0         0          0   

      Wings Joint  Women's Store  Yoga Studio  
0               0              0            0  
1               0              0            0  
2               0              0            0  
3               0              0            0  
4               0              0            0  
...           ...            ...          ...  
3874            0              0            0  
3875            0              0            0  
3876            0              0            0  
3877            0              

In [30]:
baltimore_onehot.shape

(3879, 291)

Adding the venues expanded the dataframe from (3872, 7) to (3872, 291)

We will now group the neighborhoods by taking the mean of the frequency of occurrence of each category

In [31]:
baltimore_grouped = baltimore_onehot.groupby('Location').mean().reset_index()
baltimore_grouped

Location       ATM  Adult Boutique  Afghan Restaurant  African Restaurant  \
0    Central  0.000000        0.000000           0.008000            0.002667   
1       East  0.000000        0.000000           0.000000            0.000000   
2      North  0.005176        0.001035           0.001035            0.000000   
3  Northeast  0.004202        0.000000           0.000000            0.000000   
4  Northwest  0.000000        0.000000           0.000000            0.000000   
5      South  0.000000        0.000000           0.000000            0.000000   
6  Southeast  0.001224        0.000000           0.000000            0.000000   
7  Southwest  0.000000        0.000000           0.003226            0.003226   
8       West  0.011236        0.000000           0.005618            0.005618   

    Airport  American Restaurant  Antique Shop  Aquarium  \
0  0.000000             0.029333       0.00000  0.010667   
1  0.000000             0.051724       0.00000  0.000000   
2  0.000000             0.033126       0.00207  0.000000   
3  0.000000             0.037815       0.00000  0.016807   
4  0.000000             0.052632       0.00000  0.000000   
5  0.000000             0.019943       0.00000  0.000000   
6  0.001224             0.035496       0.00000  0.000000   
7  0.000000             0.035484       0.00000  0.000000   
8  0.000000             0.056180       0.00000  0.000000   

   Argentinian Restaurant  ...  Vietnamese Restaurant  Volleyball Court  \
0                0.000000  ...               0.008000          0.000000   
1                0.000000  ...               0.005747          0.000000   
2                0.000000  ...               0.002070          0.000000   
3                0.000000  ...               0.000000          0.000000   
4                0.000000  ...               0.000000          0.000000   
5                0.000000  ...               0.000000          0.005698   
6                0.004896  ...               0.002448          0.000000   
7                0.000000  ...               0.006452          0.000000   
8                0.000000  ...               0.011236          0.000000   

   Waterfall  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0   0.000000            0.000000    0.000000  0.008000   0.000000   
1   0.000000            0.000000    0.000000  0.000000   0.000000   
2   0.001035            0.001035    0.000000  0.004141   0.004141   
3   0.000000            0.004202    0.000000  0.000000   0.004202   
4   0.000000            0.000000    0.000000  0.000000   0.000000   
5   0.000000            0.000000    0.000000  0.000000   0.000000   
6   0.000000            0.000000    0.004896  0.002448   0.003672   
7   0.003226            0.000000    0.000000  0.003226   0.000000   
8   0.000000            0.000000    0.000000  0.000000   0.000000   

   Wings Joint  Women's Store  Yoga Studio  
0     0.000000       0.000000     0.004000  
1     0.017241       0.000000     0.005747  
2     0.000000       0.003106     0.006211  
3     0.000000       0.008403     0.000000  
4     0.021053       0.000000     0.000000  
5     0.000000       0.000000     0.008547  
6     0.000000       0.008568     0.008568  
7     0.012903       0.000000     0.000000  
8     0.016854       0.000000     0.000000  

[9 rows x 291 columns]

Grouping the neighborhoods identified 9 unique values (the original geological locations), decreasing the dataframe to 9 rows and it also automatically sorted the neighborhoods

Let's identify the top 10 venues in each geological location.

In [32]:
num_top_venues = 10

for hood in baltimore_grouped['Location']:
    print("----"+hood+"----")
    temp = baltimore_grouped[baltimore_grouped['Location'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Central----
                 venue  freq
0       Sandwich Place  0.05
1                Hotel  0.04
2          Coffee Shop  0.04
3    Convenience Store  0.03
4  American Restaurant  0.03
5              Theater  0.03
6               Lounge  0.02
7                  Bar  0.02
8    Indian Restaurant  0.02
9                 Café  0.02


----East----
                  venue  freq
0   American Restaurant  0.05
1     Convenience Store  0.04
2           Coffee Shop  0.04
3              Bus Stop  0.03
4               Theater  0.03
5  Fast Food Restaurant  0.03
6    Chinese Restaurant  0.03
7        Sandwich Place  0.03
8            Shoe Store  0.02
9                  Park  0.02


----North----
                 venue  freq
0          Pizza Place  0.04
1          Coffee Shop  0.04
2                  Bar  0.04
3  American Restaurant  0.03
4       Sandwich Place  0.03
5             Pharmacy  0.03
6         Liquor Store  0.02
7   Chinese Restaurant  0.02
8             Bus Stop  0.02
9    Convenien

#### Adding the data to a pandas dataframe

In [33]:
# Function sorts the values in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [34]:
# Identifies the top 10 venues for each neighborhood and outputs them in a df
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Location']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Location'] = baltimore_grouped['Location']

for ind in np.arange(baltimore_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(baltimore_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Location 1st Most Common Venue 2nd Most Common Venue  \
0    Central        Sandwich Place                 Hotel   
1       East   American Restaurant     Convenience Store   
2      North           Coffee Shop           Pizza Place   
3  Northeast    Italian Restaurant   American Restaurant   
4  Northwest                  Park          Liquor Store   
5      South                   Bar           Pizza Place   
6  Southeast                   Bar    Italian Restaurant   
7  Southwest        Sandwich Place   American Restaurant   
8       West   American Restaurant   Fried Chicken Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop     Convenience Store               Theater   
1           Coffee Shop               Theater              Bus Stop   
2                   Bar   American Restaurant              Pharmacy   
3     Convenience Store    Seafood Restaurant    Chinese Restaurant   
4   American Restaurant    Chinese Restaurant   Arts & Crafts Store   
5        Sandwich Place                  Café      Sushi Restaurant   
6   American Restaurant    Mexican Restaurant    Seafood Restaurant   
7     Convenience Store        Discount Store        History Museum   
8        Sandwich Place           Pizza Place     Convenience Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0   American Restaurant                  Café                   Bar   
1        Sandwich Place  Fast Food Restaurant    Chinese Restaurant   
2        Sandwich Place                  Café     Convenience Store   
3   Fried Chicken Joint           Pizza Place                   Bar   
4         Grocery Store              Pharmacy        Discount Store   
5                  Park    Mexican Restaurant    Seafood Restaurant   
6           Pizza Place                 Hotel           Coffee Shop   
7           Pizza Place   Fried Chicken Joint           Coffee Shop   
8              Bus Stop                  Bank    Chinese Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0           Pizza Place      Indian Restaurant  
1          Intersection          Grocery Store  
2    Chinese Restaurant           Liquor Store  
3        Discount Store               Bus Stop  
4                Market      Convenience Store  
5                   Gym           Liquor Store  
6      Greek Restaurant         Clothing Store  
7                 Hotel                    Pub  
8                  Park          Grocery Store

## Clustering the neighborhoods

Import the necessary packages for this section

In [35]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

In [36]:
# set number of clusters
kclusters = 5

baltimore_grouped_clustering = baltimore_grouped.drop('Location', 1)
baltimore_grouped_clustering.head(10)

ATM  Adult Boutique  Afghan Restaurant  African Restaurant   Airport  \
0  0.000000        0.000000           0.008000            0.002667  0.000000   
1  0.000000        0.000000           0.000000            0.000000  0.000000   
2  0.005176        0.001035           0.001035            0.000000  0.000000   
3  0.004202        0.000000           0.000000            0.000000  0.000000   
4  0.000000        0.000000           0.000000            0.000000  0.000000   
5  0.000000        0.000000           0.000000            0.000000  0.000000   
6  0.001224        0.000000           0.000000            0.000000  0.001224   
7  0.000000        0.000000           0.003226            0.003226  0.000000   
8  0.011236        0.000000           0.005618            0.005618  0.000000   

   American Restaurant  Antique Shop  Aquarium  Argentinian Restaurant  \
0             0.029333       0.00000  0.010667                0.000000   
1             0.051724       0.00000  0.000000                0.000000   
2             0.033126       0.00207  0.000000                0.000000   
3             0.037815       0.00000  0.016807                0.000000   
4             0.052632       0.00000  0.000000                0.000000   
5             0.019943       0.00000  0.000000                0.000000   
6             0.035496       0.00000  0.000000                0.004896   
7             0.035484       0.00000  0.000000                0.000000   
8             0.056180       0.00000  0.000000                0.000000   

   Art Gallery  ...  Vietnamese Restaurant  Volleyball Court  Waterfall  \
0     0.000000  ...               0.008000          0.000000   0.000000   
1     0.011494  ...               0.005747          0.000000   0.000000   
2     0.009317  ...               0.002070          0.000000   0.001035   
3     0.000000  ...               0.000000          0.000000   0.000000   
4     0.000000  ...               0.000000          0.000000   0.000000   
5     0.000000  ...               0.000000          0.005698   0.000000   
6     0.000000  ...               0.002448          0.000000   0.000000   
7     0.000000  ...               0.006452          0.000000   0.003226   
8     0.000000  ...               0.011236          0.000000   0.000000   

   Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0            0.000000    0.000000  0.008000   0.000000     0.000000   
1            0.000000    0.000000  0.000000   0.000000     0.017241   
2            0.001035    0.000000  0.004141   0.004141     0.000000   
3            0.004202    0.000000  0.000000   0.004202     0.000000   
4            0.000000    0.000000  0.000000   0.000000     0.021053   
5            0.000000    0.000000  0.000000   0.000000     0.000000   
6            0.000000    0.004896  0.002448   0.003672     0.000000   
7            0.000000    0.000000  0.003226   0.000000     0.012903   
8            0.000000    0.000000  0.000000   0.000000     0.016854   

   Women's Store  Yoga Studio  
0       0.000000     0.004000  
1       0.000000     0.005747  
2       0.003106     0.006211  
3       0.008403     0.000000  
4       0.000000     0.000000  
5       0.000000     0.008547  
6       0.008568     0.008568  
7       0.000000     0.000000  
8       0.000000     0.000000  

[9 rows x 290 columns]

Here we are running the k-means clustering and using a random state so the results can be replicated

In [37]:
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(baltimore_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 4, 0, 3, 1, 2, 3, 0, 4], dtype=int32)

We are going to merge the new dataframe with the one we just created in the previous section. The merged dataframe should include Postal Code, Borough, Neighborhood, Lat, Long, the Clustering label just identified, and the top 7 venues for each neighborhood.

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [39]:
baltimore_merged = df_final

# merge baltimore_grouped with df_final to add latitude/longitude for each neighborhood
baltimore_merged = baltimore_merged.join(neighborhoods_venues_sorted.set_index('Location'), on='Location')

In [40]:
baltimore_merged.head()

Neighborhood   Location   Latitude  Longitude  Cluster Labels  \
0              Abell      North  39.326139 -76.611737               0   
1          Allendale       West  39.291687 -76.676942               4   
2            Arcadia  Northeast  39.334056 -76.566500               3   
3          Arlington  Northwest  39.348692 -76.682666               1   
4  Armistead Gardens  Northeast  39.306646 -76.556279               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop           Pizza Place                   Bar   
1   American Restaurant   Fried Chicken Joint        Sandwich Place   
2    Italian Restaurant   American Restaurant     Convenience Store   
3                  Park          Liquor Store   American Restaurant   
4    Italian Restaurant   American Restaurant     Convenience Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0   American Restaurant              Pharmacy        Sandwich Place   
1           Pizza Place     Convenience Store              Bus Stop   
2    Seafood Restaurant    Chinese Restaurant   Fried Chicken Joint   
3    Chinese Restaurant   Arts & Crafts Store         Grocery Store   
4    Seafood Restaurant    Chinese Restaurant   Fried Chicken Joint   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                  Café     Convenience Store    Chinese Restaurant   
1                  Bank    Chinese Restaurant                  Park   
2           Pizza Place                   Bar        Discount Store   
3              Pharmacy        Discount Store                Market   
4           Pizza Place                   Bar        Discount Store   

  10th Most Common Venue  
0           Liquor Store  
1          Grocery Store  
2               Bus Stop  
3      Convenience Store  
4               Bus Stop

We will now put all the values on a map of Baltimore

In [41]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(baltimore_merged['Latitude'], baltimore_merged['Longitude'], baltimore_merged['Neighborhood'], baltimore_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examining the clusters created

First cluster: Largest Cluster- Most Diverse

In [42]:
baltimore_merged.loc[baltimore_merged['Cluster Labels'] == 0, baltimore_merged.columns[[0] + list(range(4, baltimore_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
0                 Abell               0           Coffee Shop   
9          Barre Circle               0        Sandwich Place   
11           Beechfield               0        Sandwich Place   
17          Bolton Hill               0        Sandwich Place   
24      Cameron Village               0           Coffee Shop   
..                  ...             ...                   ...   
213  Woodbourne Heights               0           Coffee Shop   
214   Woodbourne-McCabe               0           Coffee Shop   
217          Wyman Park               0           Coffee Shop   
218           Wyndhurst               0           Coffee Shop   
219        Yale Heights               0        Sandwich Place   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0             Pizza Place                   Bar   American Restaurant   
9     American Restaurant     Convenience Store        Discount Store   
11    American Restaurant     Convenience Store        Discount Store   
17                  Hotel           Coffee Shop     Convenience Store   
24            Pizza Place                   Bar   American Restaurant   
..                    ...                   ...                   ...   
213           Pizza Place                   Bar   American Restaurant   
214           Pizza Place                   Bar   American Restaurant   
217           Pizza Place                   Bar   American Restaurant   
218           Pizza Place                   Bar   American Restaurant   
219   American Restaurant     Convenience Store        Discount Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Pharmacy        Sandwich Place                  Café   
9          History Museum           Pizza Place   Fried Chicken Joint   
11         History Museum           Pizza Place   Fried Chicken Joint   
17                Theater   American Restaurant                  Café   
24               Pharmacy        Sandwich Place                  Café   
..                    ...                   ...                   ...   
213              Pharmacy        Sandwich Place                  Café   
214              Pharmacy        Sandwich Place                  Café   
217              Pharmacy        Sandwich Place                  Café   
218              Pharmacy        Sandwich Place                  Café   
219        History Museum           Pizza Place   Fried Chicken Joint   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0       Convenience Store    Chinese Restaurant           Liquor Store  
9             Coffee Shop                 Hotel                    Pub  
11            Coffee Shop                 Hotel                    Pub  
17                    Bar           Pizza Place      Indian Restaurant  
24      Convenience Store    Chinese Restaurant           Liquor Store  
..                    ...                   ...                    ...  
213     Convenience Store    Chinese Restaurant           Liquor Store  
214     Convenience Store    Chinese Restaurant           Liquor Store  
217     Convenience Store    Chinese Restaurant           Liquor Store  
218     Convenience Store    Chinese Restaurant           Liquor Store  
219           Coffee Shop                 Hotel                    Pub  

[88 rows x 12 columns]

Second cluster: Park/Drinking

In [43]:
baltimore_merged.loc[baltimore_merged['Cluster Labels'] == 1, baltimore_merged.columns[[0] + list(range(4, baltimore_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
3               Arlington               1                  Park   
6               Ashburton               1                  Park   
23      Callaway-Garrison               1                  Park   
31   Central Park Heights               1                  Park   
36              Cheswolde               1                  Park   
40          Cross Country               1                  Park   
45               Dolfield               1                  Park   
46             Dorchester               1                  Park   
49         East Arlington               1                  Park   
59              Fallstaff               1                  Park   
63            Forest Park               1                  Park   
69                   Glen               1                  Park   
77             Grove Park               1                  Park   
79              Gwynn Oak               1                  Park   
81            Hanlon Park               1                  Park   
96            Howard Park               1                  Park   
110       Langston Hughes               1                  Park   
115        Liberty Square               1                  Park   
119          Lucille Park               1                  Park   
150           Park Circle               1                  Park   
151          Park Heights               1                  Park   
160               Pimlico               1                  Park   
165               Purnell               1                  Park   
202      West Forest Park               1                  Park   
203      West Hill Square               1                  Park   
210         Windsor Hills               1                  Park   
215              Woodmoor               1                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3            Liquor Store   American Restaurant    Chinese Restaurant   
6            Liquor Store   American Restaurant    Chinese Restaurant   
23           Liquor Store   American Restaurant    Chinese Restaurant   
31           Liquor Store   American Restaurant    Chinese Restaurant   
36           Liquor Store   American Restaurant    Chinese Restaurant   
40           Liquor Store   American Restaurant    Chinese Restaurant   
45           Liquor Store   American Restaurant    Chinese Restaurant   
46           Liquor Store   American Restaurant    Chinese Restaurant   
49           Liquor Store   American Restaurant    Chinese Restaurant   
59           Liquor Store   American Restaurant    Chinese Restaurant   
63           Liquor Store   American Restaurant    Chinese Restaurant   
69           Liquor Store   American Restaurant    Chinese Restaurant   
77           Liquor Store   American Restaurant    Chinese Restaurant   
79           Liquor Store   American Restaurant    Chinese Restaurant   
81           Liquor Store   American Restaurant    Chinese Restaurant   
96           Liquor Store   American Restaurant    Chinese Restaurant   
110          Liquor Store   American Restaurant    Chinese Restaurant   
115          Liquor Store   American Restaurant    Chinese Restaurant   
119          Liquor Store   American Restaurant    Chinese Restaurant   
150          Liquor Store   American Restaurant    Chinese Restaurant   
151          Liquor Store   American Restaurant    Chinese Restaurant   
160          Liquor Store   American Restaurant    Chinese Restaurant   
165          Liquor Store   American Restaurant    Chinese Restaurant   
202          Liquor Store   American Restaurant    Chinese Restaurant   
203          Liquor Store   American Restaurant    Chinese Restaurant   
210          Liquor Store   American Restaurant    Chinese Restaurant   
215          Liquor Store   American Restaurant    Chinese Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3     Arts & Craft

Third cluster: Social groups

In [44]:
baltimore_merged.loc[baltimore_merged['Cluster Labels'] == 2, baltimore_merged.columns[[0] + list(range(4, baltimore_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
5         Arundel Cove               2                   Bar   
20            Brooklyn               2                   Bar   
21      Brooklyn Homes               2                   Bar   
35         Cherry Hill               2                   Bar   
41          Curtis Bay               2                   Bar   
60        Federal Hill               2                   Bar   
84          Harborview               2                   Bar   
87       Hawkins Point               2                   Bar   
109           Lakeland               2                   Bar   
117       Locust Point               2                   Bar   
122         Masonville               2                   Bar   
136       Mount Winans               2                   Bar   
162     Port Covington               2                   Bar   
171          Riverside               2                   Bar   
180   Sharp-Leadenhall               2                   Bar   
195     Wagner's Point               2                   Bar   
201  West Federal Hill               2                   Bar   
208           Westport               2                   Bar   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5             Pizza Place        Sandwich Place                  Café   
20            Pizza Place        Sandwich Place                  Café   
21            Pizza Place        Sandwich Place                  Café   
35            Pizza Place        Sandwich Place                  Café   
41            Pizza Place        Sandwich Place                  Café   
60            Pizza Place        Sandwich Place                  Café   
84            Pizza Place        Sandwich Place                  Café   
87            Pizza Place        Sandwich Place                  Café   
109           Pizza Place        Sandwich Place                  Café   
117           Pizza Place        Sandwich Place                  Café   
122           Pizza Place        Sandwich Place                  Café   
136           Pizza Place        Sandwich Place                  Café   
162           Pizza Place        Sandwich Place                  Café   
171           Pizza Place        Sandwich Place                  Café   
180           Pizza Place        Sandwich Place                  Café   
195           Pizza Place        Sandwich Place                  Café   
201           Pizza Place        Sandwich Place                  Café   
208           Pizza Place        Sandwich Place                  Café   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5        Sushi Restaurant                  Park    Mexican Restaurant   
20       Sushi Restaurant                  Park    Mexican Restaurant   
21       Sushi Restaurant                  Park    Mexican Restaurant   
35       Sushi Restaurant                  Park    Mexican Restaurant   
41       Sushi Restaurant                  Park    Mexican Restaurant   
60       Sushi Restaurant                  Park    Mexican Restaurant   
84       Sushi Restaurant                  Park    Mexican Restaurant   
87       Sushi Restaurant                  Park    Mexican Restaurant   
109      Sushi Restaurant                  Park    Mexican Restaurant   
117      Sushi Restaurant                  Park    Mexican Restaurant   
122      Sushi Restaurant                  Park    Mexican Restaurant   
136      Sushi Restaurant                  Park    Mexican Restaurant   
162      Sushi Restaurant                  Park    Mexican Restaurant   
171      Sushi Restaurant                  Park    Mexican Restaurant   
180      Sushi Restaurant                  Park    Mexican Restaurant   
195      Sushi Restaurant                  Park    Mexican Restaurant   
201      Sushi Restaurant                  Park    Mexican Restaurant   
208      Sushi Restaurant                  Park    Mexican Restaurant   

    8th Most Comm

Fourth Cluster: Restaurants

In [45]:
baltimore_merged.loc[baltimore_merged['Cluster Labels'] == 3, baltimore_merged.columns[[0] + list(range(4, baltimore_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
2                   Arcadia               3    Italian Restaurant   
4         Armistead Gardens               3    Italian Restaurant   
7       Baltimore Highlands               3                   Bar   
10                  Bayview               3                   Bar   
12            Belair-Edison               3    Italian Restaurant   
13          Belair-Parkside               3    Italian Restaurant   
15            Beverly Hills               3    Italian Restaurant   
18            Brewer's Hill               3                   Bar   
22            Butchers Hill               3                   Bar   
25                   Canton               3                   Bar   
29                  Cedmont               3    Italian Restaurant   
30                  Cedonia               3    Italian Restaurant   
37              Christopher               3    Italian Restaurant   
52                 Eastwood               3                   Bar   
55   Ednor Gardens-Lakeside               3    Italian Restaurant   
61              Fells Point               3                   Bar   
62           Fells Prospect               3                   Bar   
64                Frankford               3    Italian Restaurant   
67              Gardenville               3    Italian Restaurant   
71          Glenham-Belford               3    Italian Restaurant   
73           Graceland Park               3                   Bar   
74                Greektown               3                   Bar   
82              Harbor East               3                   Bar   
83             Harbor Point               3                   Bar   
89             Highlandtown               3                   Bar   
90                   Hillen               3    Italian Restaurant   
93          Hollander Ridge               3    Italian Restaurant   
97                 Idlewood               3    Italian Restaurant   
101               Jonestown               3                   Bar   
102              Joseph Lee               3                   Bar   
106                 Kresson               3                   Bar   
111              Lauraville               3    Italian Restaurant   
116            Little Italy               3                   Bar   
123                Mayfield               3    Italian Restaurant   
130             Morgan Park               3    Italian Restaurant   
137      North Harford Road               3    Italian Restaurant   
138               Northwood               3    Italian Restaurant   
141       O'Donnell Heights               3                   Bar   
146           Orchard Ridge               3    Italian Restaurant   
148                 Overlea               3    Italian Restaurant   
152                Parkside               3    Italian Restaurant   
153          Patterson Park               3                   Bar   
157           Perkins Homes               3                   Bar   
158            Perring-Loch               3    Italian Restaurant   
167              Ramblewood               3    Italian Restaurant   
176         Rosemont Avenue               3    Italian Restaurant   
182               St Helena               3                   Bar   
184          Taylor Heights               3    Italian Restaurant   
191       Upper Fells Point               3                   Bar   
198              Waltherson               3    Italian Restaurant   
199         Washington Hill               3                   Bar   
206               Westfield               3    Italian Restaurant   
216                Woodring               3    Italian Restaurant   
220               York Road               3    Italian Restaurant   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2     American Restaurant     Convenience Store    Seafood Restaurant   
4     American Restaurant     Convenience Store    Seafood Restaurant   


Fifth Cluster: Quick bites 

In [46]:
baltimore_merged.loc[baltimore_merged['Cluster Labels'] == 4, baltimore_merged.columns[[0] + list(range(4, baltimore_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
1                Allendale               4   American Restaurant   
8                  Barclay               4   American Restaurant   
14                   Berea               4   American Restaurant   
16           Biddle Street               4   American Restaurant   
19           Broadway East               4   American Restaurant   
39          Coppin Heights               4   American Restaurant   
43             Darley Park               4   American Restaurant   
48              Eager Park               4   American Restaurant   
50   East Baltimore Midway               4   American Restaurant   
51              Easterwood               4   American Restaurant   
54       Edmondson Village               4   American Restaurant   
56            Ellwood Park               4   American Restaurant   
65         Franklin Square               4   American Restaurant   
68              Gay Street               4   American Restaurant   
75         Greenmount West               4   American Restaurant   
85             Harlem Park               4   American Restaurant   
88       Heritage Crossing               4   American Restaurant   
100        Johnston Square               4   American Restaurant   
113              Lexington               4   American Restaurant   
124         McElderry Park               4   American Restaurant   
129              Mondawmin               4   American Restaurant   
132                 Mosher               4   American Restaurant   
143               Old Town               4   American Restaurant   
144                 Oliver               4   American Restaurant   
145            Orangeville               4   American Restaurant   
149        Panway-Braddish               4   American Restaurant   
155          Penn-Fallsway               4   American Restaurant   
156             Penn-North               4   American Restaurant   
161              Poppleton               4   American Restaurant   
175               Rosemont               4   American Restaurant   
178    Sandtown-Winchester               4   American Restaurant   
188           Union Square               4   American Restaurant   
192                  Upton               4   American Restaurant   
197               Walbrook               4   American Restaurant   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1     Fried Chicken Joint        Sandwich Place           Pizza Place   
8       Convenience Store           Coffee Shop               Theater   
14      Convenience Store           Coffee Shop               Theater   
16      Convenience Store           Coffee Shop               Theater   
19      Convenience Store           Coffee Shop               Theater   
39    Fried Chicken Joint        Sandwich Place           Pizza Place   
43      Convenience Store           Coffee Shop               Theater   
48      Convenience Store           Coffee Shop               Theater   
50      Convenience Store           Coffee Shop               Theater   
51    Fried Chicken Joint        Sandwich Place           Pizza Place   
54    Fried Chicken Joint        Sandwich Place           Pizza Place   
56      Convenience Store           Coffee Shop               Theater   
65    Fried Chicken Joint        Sandwich Place           Pizza Place   
68      Convenience Store           Coffee Shop               Theater   
75      Convenience Store           Coffee Shop               Theater   
85    Fried Chicken Joint        Sandwich Place           Pizza Place   
88    Fried Chicken Joint        Sandwich Place           Pizza Place   
100     Convenience Store           Coffee Shop               Theater   
113   Fried Chicken Joint        Sandwich Place           Pizza Place   
124     Convenience Store           Coffee Shop               Theater   
129   Fried Chicken Joint        Sandwich Place           Pizza Place   
132   Fried Chicken Joint  

### Examination by geological location
After using k-means on the geological location, the clusters were grouped similar to the location. We will now look complete the same analysis, but this time we will be using the neighborhood names instead to see if the grouping is similar or different.

### Start of Neighborhood Analysis

In [47]:
def getNearbyVenues2(names, latitudes, longitudes, radius=500):
    
    venues_list2=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list2.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues2 = pd.DataFrame([item for venue_list2 in venues_list2 for item in venue_list2])
    nearby_venues2.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues2)

In [48]:
baltimore2_venues = getNearbyVenues2(names=df_final['Neighborhood'],
                                   latitudes=df_final['Latitude'],
                                   longitudes=df_final['Longitude']
                                  )

Abell
Allendale
Arcadia
Arlington
Armistead Gardens
Arundel Cove
Ashburton
Baltimore Highlands
Barclay
Barre Circle
Bayview
Beechfield
Belair-Edison
Belair-Parkside
Berea
Beverly Hills
Biddle Street
Bolton Hill
Brewer's Hill
Broadway East
Brooklyn
Brooklyn Homes
Butchers Hill
Callaway-Garrison
Cameron Village
Canton
Carrollton Ridge
Cathedral Hill
Cedarcroft
Cedmont
Cedonia
Central Park Heights
Charles Center
Charles North
Charles Village
Cherry Hill
Cheswolde
Christopher
Cold Spring
Coppin Heights
Cross Country
Curtis Bay
Cylburn
Darley Park
Dickeyville
Dolfield
Dorchester
Druid Heights
Eager Park
East Arlington
East Baltimore Midway
Easterwood
Eastwood
Edmondson
Edmondson Village
Ednor Gardens-Lakeside
Ellwood Park
Evergreen
Fairmount
Fallstaff
Federal Hill
Fells Point
Fells Prospect
Forest Park
Frankford
Franklin Square
Franklintown
Gardenville
Gay Street
Glen
Glen Oaks
Glenham-Belford
Govans
Graceland Park
Greektown
Greenmount West
Greenspring
Grove Park
Guilford
Gwynn Oak
Hampden


In [49]:
print(baltimore2_venues.shape)
baltimore2_venues.head()

(3879, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0        Abell              39.326139              -76.611737   
1        Abell              39.326139              -76.611737   
2        Abell              39.326139              -76.611737   
3        Abell              39.326139              -76.611737   
4        Abell              39.326139              -76.611737   

                         Venue  Venue Latitude  Venue Longitude  \
0  32nd Street Farmer's Market       39.327362       -76.610851   
1                Pete's Grille       39.327064       -76.609593   
2               The Book Thing       39.325253       -76.610272   
3     Normal's Books & Records       39.326012       -76.609903   
4      Peabody Heights Brewery       39.324744       -76.610819   

   Venue Category  
0  Farmers Market  
1  Breakfast Spot  
2       Bookstore  
3     Record Shop  
4         Brewery

In [50]:
print(baltimore2_venues.shape)
baltimore2_venues.head()

(3879, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0        Abell              39.326139              -76.611737   
1        Abell              39.326139              -76.611737   
2        Abell              39.326139              -76.611737   
3        Abell              39.326139              -76.611737   
4        Abell              39.326139              -76.611737   

                         Venue  Venue Latitude  Venue Longitude  \
0  32nd Street Farmer's Market       39.327362       -76.610851   
1                Pete's Grille       39.327064       -76.609593   
2               The Book Thing       39.325253       -76.610272   
3     Normal's Books & Records       39.326012       -76.609903   
4      Peabody Heights Brewery       39.324744       -76.610819   

   Venue Category  
0  Farmers Market  
1  Breakfast Spot  
2       Bookstore  
3     Record Shop  
4         Brewery

In [51]:
# one hot encoding
baltimore2_onehot = pd.get_dummies(baltimore2_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
baltimore2_onehot['Neighborhood'] = baltimore2_venues['Neighborhood']

In [52]:
# move neighborhood column to the first column
neighborhood = baltimore2_onehot['Neighborhood']
baltimore2_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
baltimore2_onehot.insert(0, 'Neighborhood', neighborhood)
baltimore2_onehot

Neighborhood  ATM  Adult Boutique  Afghan Restaurant  African Restaurant  \
0           Abell    0               0                  0                   0   
1           Abell    0               0                  0                   0   
2           Abell    0               0                  0                   0   
3           Abell    0               0                  0                   0   
4           Abell    0               0                  0                   0   
...           ...  ...             ...                ...                 ...   
3874    York Road    0               0                  0                   0   
3875    York Road    0               0                  0                   0   
3876    York Road    0               0                  0                   0   
3877    York Road    0               0                  0                   0   
3878    York Road    0               0                  0                   0   

      Airport  American Restaurant  Antique Shop  Aquarium  \
0           0                    0             0         0   
1           0                    0             0         0   
2           0                    0             0         0   
3           0                    0             0         0   
4           0                    0             0         0   
...       ...                  ...           ...       ...   
3874        0                    0             0         0   
3875        0                    0             0         0   
3876        0                    0             0         0   
3877        0                    0             0         0   
3878        0                    0             0         0   

      Argentinian Restaurant  ...  Vietnamese Restaurant  Volleyball Court  \
0                          0  ...                      0                 0   
1                          0  ...                      0                 0   
2                          0  ...                      0                 0   
3                          0  ...                      0                 0   
4                          0  ...                      0                 0   
...                      ...  ...                    ...               ...   
3874                       0  ...                      0                 0   
3875                       0  ...                      0                 0   
3876                       0  ...                      0                 0   
3877                       0  ...                      0                 0   
3878                       0  ...                      0                 0   

      Waterfall  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0             0                   0           0         0          0   
1             0                   0           0         0          0   
2             0                   0           0         0          0   
3             0                   0           0         0          0   
4             0                   0           0         0          0   
...         ...                 ...         ...       ...        ...   
3874          0                   0           0         0          0   
3875          0                   0           0         0          0   
3876          0                   0           0         0          0   
3877          0                   0           0         0          0   
3878          0                   0           0         0          0   

      Wings Joint  Women's Store  Yoga Studio  
0               0              0            0  
1               0              0            0  
2               0              0            0  
3               0              0            0  
4               0              0            0  
...           ...            ...          ...  
3874            0              0            0  
3875            0              0            0  
3876            0              0            0  
3877 

In [53]:
baltimore2_onehot.shape

(3879, 290)

In [54]:
baltimore2_grouped = baltimore2_onehot.groupby('Neighborhood').mean().reset_index()
baltimore2_grouped

Neighborhood  ATM  Adult Boutique  Afghan Restaurant  \
0                Abell  0.0             0.0                0.0   
1            Allendale  0.0             0.0                0.0   
2              Arcadia  0.0             0.0                0.0   
3            Arlington  0.0             0.0                0.0   
4    Armistead Gardens  0.0             0.0                0.0   
..                 ...  ...             ...                ...   
207           Woodring  0.0             0.0                0.0   
208         Wyman Park  0.0             0.0                0.0   
209          Wyndhurst  0.0             0.0                0.0   
210       Yale Heights  0.0             0.0                0.0   
211          York Road  0.0             0.0                0.0   

     African Restaurant  Airport  American Restaurant  Antique Shop  Aquarium  \
0                   0.0      0.0             0.000000      0.000000       0.0   
1                   0.0      0.0             0.333333      0.000000       0.0   
2                   0.0      0.0             0.000000      0.000000       0.0   
3                   0.0      0.0             0.000000      0.000000       0.0   
4                   0.0      0.0             0.000000      0.000000       0.0   
..                  ...      ...                  ...           ...       ...   
207                 0.0      0.0             0.000000      0.000000       0.0   
208                 0.0      0.0             0.000000      0.023256       0.0   
209                 0.0      0.0             0.000000      0.000000       0.0   
210                 0.0      0.0             0.000000      0.000000       0.0   
211                 0.0      0.0             0.000000      0.000000       0.0   

     Argentinian Restaurant  ...  Vietnamese Restaurant  Volleyball Court  \
0                       0.0  ...                    0.0               0.0   
1                       0.0  ...                    0.0               0.0   
2                       0.0  ...                    0.0               0.0   
3                       0.0  ...                    0.0               0.0   
4                       0.0  ...                    0.0               0.0   
..                      ...  ...                    ...               ...   
207                     0.0  ...                    0.0               0.0   
208                     0.0  ...                    0.0               0.0   
209                     0.0  ...                    0.0               0.0   
210                     0.0  ...                    0.0               0.0   
211                     0.0  ...                    0.0               0.0   

     Waterfall  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0          0.0                 0.0         0.0       0.0   0.000000   
1          0.0                 0.0         0.0       0.0   0.000000   
2          0.0                 0.0         0.0       0.0   0.000000   
3          0.0                 0.0         0.0       0.0   0.000000   
4          0.0                 0.0         0.0       0.0   0.000000   
..         ...                 ...         ...       ...        ...   
207        0.0                 0.0         0.0       0.0   0.000000   
208        0.0                 0.0         0.0       0.0   0.023256   
209        0.0                 0.0         0.0       0.0   0.000000   
210        0.0                 0.0         0.0       0.0   0.000000   
211        0.0                 0.0         0.0       0.0   0.000000   

     Wings Joint  Women's Store  Yoga Studio  
0            0.0       0.000000     0.000000  
1            0.0       0.000000     0.000000  
2            0.0       0.000000     0.000000  
3            0.0       0.000000     0.000000  
4            0.0       0.000000     0.000000  
..           ...            ...          ...  
207          0.0       0.000000     0.000000  
208          0.0       0.000000     0.023256  
209          0.0       0.090909     0.00000

In [55]:
num_top_venues = 10

for hood in baltimore2_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = baltimore2_grouped[baltimore2_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Abell----
                venue  freq
0         Pizza Place  0.07
1  Chinese Restaurant  0.07
2            Pharmacy  0.05
3      Sandwich Place  0.05
4         Coffee Shop  0.04
5       Grocery Store  0.04
6            Bus Stop  0.04
7                 Bar  0.04
8                Café  0.04
9           Bookstore  0.04


----Allendale----
                   venue  freq
0                    Bar  0.33
1    American Restaurant  0.33
2       Basketball Court  0.33
3                    ATM  0.00
4        Organic Grocery  0.00
5     Persian Restaurant  0.00
6  Performing Arts Venue  0.00
7                   Park  0.00
8      Outdoor Sculpture  0.00
9      Other Repair Shop  0.00


----Arcadia----
                   venue  freq
0     Chinese Restaurant   0.5
1      Other Repair Shop   0.5
2                    ATM   0.0
3              Pet Store   0.0
4    Peruvian Restaurant   0.0
5     Persian Restaurant   0.0
6  Performing Arts Venue   0.0
7                   Park   0.0
8      Outdoor Sculp

In [56]:
# Function sorts the values in descending order
def return_most_common_venues2(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [70]:
# Identifies the top 10 venues for each neighborhood and outputs them in a df
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted3 = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted3['Neighborhood'] = baltimore2_grouped['Neighborhood']

for ind in np.arange(baltimore2_grouped.shape[0]):
    neighborhoods_venues_sorted3.iloc[ind, 1:] = return_most_common_venues2(baltimore2_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted3

Neighborhood 1st Most Common Venue   2nd Most Common Venue  \
0                Abell           Pizza Place      Chinese Restaurant   
1            Allendale                   Bar     American Restaurant   
2              Arcadia    Chinese Restaurant       Other Repair Shop   
3            Arlington  Caribbean Restaurant          Discount Store   
4    Armistead Gardens     Convenience Store          Rental Service   
..                 ...                   ...                     ...   
207           Woodring        Baseball Field       German Restaurant   
208         Wyman Park                Bakery             Pizza Place   
209          Wyndhurst                  Bank                     Spa   
210       Yale Heights          Home Service             Yoga Studio   
211          York Road  Fast Food Restaurant  Thrift / Vintage Store   

               3rd Most Common Venue   4th Most Common Venue  \
0                     Sandwich Place                Pharmacy   
1                   Basketball Court             Yoga Studio   
2                        Yoga Studio       Electronics Store   
3    Southern / Soul Food Restaurant  Thrift / Vintage Store   
4                     Sandwich Place                     Bar   
..                               ...                     ...   
207                      Yoga Studio    Ethiopian Restaurant   
208                        Pet Store       Indian Restaurant   
209                      Coffee Shop             Flower Shop   
210              Dumpling Restaurant    Ethiopian Restaurant   
211                 Business Service     Fried Chicken Joint   

      5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Bus Stop           Coffee Shop                  Café   
1                     Field               Exhibit    Falafel Restaurant   
2             Event Service           Event Space               Exhibit   
3     Check Cashing Service    Chinese Restaurant              Pharmacy   
4        Seafood Restaurant           Fish Market                 Field   
..                      ...                   ...                   ...   
207             Event Space               Exhibit    Falafel Restaurant   
208                    Café           Coffee Shop           Yoga Studio   
209  Furniture / Home Store      Toy / Game Store              Pharmacy   
210           Event Service           Event Space               Exhibit   
211             Coffee Shop        Farmers Market         Event Service   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Grocery Store     Convenience Store              Bookstore  
1                    Farm        Farmers Market   Fast Food Restaurant  
2      Falafel Restaurant                  Farm         Farmers Market  
3       Convenience Store      Department Store          Design Studio  
4             Flea Market  Fast Food Restaurant         Farmers Market  
..                    ...                   ...                    ...  
207                  Farm        Farmers Market   Fast Food Restaurant  
208             Bookstore         Shopping Mall             Shoe Store  
209        Clothing Store         Grocery Store          Women's Store  
210    Falafel Restaurant                  Farm         Farmers Market  
211           Event Space               Exhibit     Falafel Restaurant  

[212 rows x 11 columns]

# Clustering Baltimore Neighborhoods

In [71]:
# set number of clusters
kclusters = 5

baltimore2_grouped_clustering = baltimore2_grouped.drop('Neighborhood', 1)
baltimore2_grouped_clustering.head(10)

ATM  Adult Boutique  Afghan Restaurant  African Restaurant  Airport  \
0  0.0             0.0                0.0                 0.0     0.00   
1  0.0             0.0                0.0                 0.0     0.00   
2  0.0             0.0                0.0                 0.0     0.00   
3  0.0             0.0                0.0                 0.0     0.00   
4  0.0             0.0                0.0                 0.0     0.00   
5  0.0             0.0                0.0                 0.0     0.00   
6  0.0             0.0                0.0                 0.0     0.00   
7  0.0             0.0                0.0                 0.0     0.25   
8  0.0             0.0                0.0                 0.0     0.00   
9  0.0             0.0                0.0                 0.0     0.00   

   American Restaurant  Antique Shop  Aquarium  Argentinian Restaurant  \
0             0.000000           0.0       0.0                     0.0   
1             0.333333           0.0       0.0                     0.0   
2             0.000000           0.0       0.0                     0.0   
3             0.000000           0.0       0.0                     0.0   
4             0.000000           0.0       0.0                     0.0   
5             0.000000           0.0       0.0                     0.0   
6             0.000000           0.0       0.0                     0.0   
7             0.000000           0.0       0.0                     0.0   
8             0.150000           0.0       0.0                     0.0   
9             0.062500           0.0       0.0                     0.0   

   Art Gallery  ...  Vietnamese Restaurant  Volleyball Court  Waterfall  \
0          0.0  ...                    0.0               0.0        0.0   
1          0.0  ...                    0.0               0.0        0.0   
2          0.0  ...                    0.0               0.0        0.0   
3          0.0  ...                    0.0               0.0        0.0   
4          0.0  ...                    0.0               0.0        0.0   
5          0.0  ...                    0.0               0.0        0.0   
6          0.0  ...                    0.0               0.0        0.0   
7          0.0  ...                    0.0               0.0        0.0   
8          0.0  ...                    0.0               0.0        0.0   
9          0.0  ...                    0.0               0.0        0.0   

   Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                 0.0         0.0       0.0        0.0          0.0   
1                 0.0         0.0       0.0        0.0          0.0   
2                 0.0         0.0       0.0        0.0          0.0   
3                 0.0         0.0       0.0        0.0          0.0   
4                 0.0         0.0       0.0        0.0          0.0   
5                 0.0         0.0       0.0        0.0          0.0   
6                 0.0         0.0       0.0        0.0          0.0   
7                 0.0         0.0       0.0        0.0          0.0   
8                 0.0         0.0       0.0        0.0          0.0   
9                 0.0         0.0       0.0        0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  
5            0.0          0.0  
6            0.0          0.0  
7            0.0          0.0  
8            0.0          0.0  
9            0.0          0.0  

[10 rows x 289 columns]

In [72]:
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(baltimore2_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 0, 0, 2, 0, 2, 2], dtype=int32)

In [73]:
# add clustering labels
neighborhoods_venues_sorted3.insert(0, 'Cluster Labels', kmeans.labels_)

In [74]:
baltimore2_merged = df_final

# merge baltimore_grouped with df_final to add latitude/longitude for each neighborhood
baltimore2_merged = baltimore2_merged.join(neighborhoods_venues_sorted3.set_index('Neighborhood'), on='Neighborhood')

baltimore2_merged.head()


Neighborhood   Location   Latitude  Longitude  Cluster Labels  \
0              Abell      North  39.326139 -76.611737             2.0   
1          Allendale       West  39.291687 -76.676942             2.0   
2            Arcadia  Northeast  39.334056 -76.566500             2.0   
3          Arlington  Northwest  39.348692 -76.682666             2.0   
4  Armistead Gardens  Northeast  39.306646 -76.556279             0.0   

  1st Most Common Venue 2nd Most Common Venue  \
0           Pizza Place    Chinese Restaurant   
1                   Bar   American Restaurant   
2    Chinese Restaurant     Other Repair Shop   
3  Caribbean Restaurant        Discount Store   
4     Convenience Store        Rental Service   

             3rd Most Common Venue   4th Most Common Venue  \
0                   Sandwich Place                Pharmacy   
1                 Basketball Court             Yoga Studio   
2                      Yoga Studio       Electronics Store   
3  Southern / Soul Food Restaurant  Thrift / Vintage Store   
4                   Sandwich Place                     Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0               Bus Stop           Coffee Shop                  Café   
1                  Field               Exhibit    Falafel Restaurant   
2          Event Service           Event Space               Exhibit   
3  Check Cashing Service    Chinese Restaurant              Pharmacy   
4     Seafood Restaurant           Fish Market                 Field   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Grocery Store     Convenience Store              Bookstore  
1                  Farm        Farmers Market   Fast Food Restaurant  
2    Falafel Restaurant                  Farm         Farmers Market  
3     Convenience Store      Department Store          Design Studio  
4           Flea Market  Fast Food Restaurant         Farmers Market

In [75]:
baltimore2_merged.loc[baltimore2_merged['Cluster Labels'] == 0, baltimore2_merged.columns[[0] + list(range(4, baltimore2_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
4      Armistead Gardens             0.0     Convenience Store   
5           Arundel Cove             0.0                   Gym   
7    Baltimore Highlands             0.0     Convenience Store   
13       Belair-Parkside             0.0     Convenience Store   
14                 Berea             0.0           Wings Joint   
35           Cherry Hill             0.0          Dessert Shop   
45              Dolfield             0.0     Convenience Store   
64             Frankford             0.0     Convenience Store   
71       Glenham-Belford             0.0     Convenience Store   
90                Hillen             0.0      Sculpture Garden   
102           Joseph Lee             0.0     Convenience Store   
115       Liberty Square             0.0     Convenience Store   
121        Market Center             0.0                   Gym   
149      Panway-Braddish             0.0     Convenience Store   
158         Perring-Loch             0.0                   Gym   
178  Sandtown-Winchester             0.0                Bakery   
183        Station North             0.0                   Gym   
198           Waltherson             0.0     Convenience Store   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4          Rental Service        Sandwich Place                   Bar   
5         Harbor / Marina     Convenience Store        Baseball Field   
7                     Bar               Airport    Light Rail Station   
13                  River    Chinese Restaurant                Lounge   
14               Bus Stop                Market     Convenience Store   
35      Convenience Store               Dog Run           Auto Garage   
45          Grocery Store   Arts & Crafts Store        Farmers Market   
64      Indian Restaurant         Shopping Mall    Chinese Restaurant   
71           Intersection  Ethiopian Restaurant         Event Service   
90               Bus Stop   Monument / Landmark     Convenience Store   
102          Optical Shop          Liquor Store                  Park   
115   American Restaurant    Chinese Restaurant            Restaurant   
121       Harbor / Marina                  Food                  Park   
149        Sandwich Place  Fast Food Restaurant             Locksmith   
158            Food Stand     Electronics Store  Ethiopian Restaurant   
178     Convenience Store   Fried Chicken Joint                 Plaza   
183       Harbor / Marina                  Food                  Park   
198                   Gym     Food & Drink Shop                  Food   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4      Seafood Restaurant           Fish Market                 Field   
5              Food Court          Dance Studio         Event Service   
7          Farmers Market         Event Service           Event Space   
13            Flower Shop           Flea Market           Fish Market   
14                   Food        Farmers Market         Event Service   
35     Light Rail Station              Dive Bar  Fast Food Restaurant   
45          Event Service           Event Space               Exhibit   
64         Farmers Market         Event Service           Event Space   
71            Event Space               Exhibit    Falafel Restaurant   
90           Cupcake Shop         Event Service     Food & Drink Shop   
102    Chinese Restaurant        Baseball Field            Restaurant   
115          Intersection                  Food                  Farm   
121              Pharmacy     Convenience Store         Deli / Bodega   
149      Cuban Restaurant     Electronics Store     Food & Drink Shop   
158         Event Service           Event Space               Exhibit   
178  Fast Food Restaurant         Event Service           Event Space   
183              Pharmacy     Convenience Store         Deli / Bodega   
198           Flower Shop           Flea Market        

In [76]:
baltimore2_merged.loc[baltimore2_merged['Cluster Labels'] == 1, baltimore2_merged.columns[[0] + list(range(4, baltimore2_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
19          Broadway East             1.0                  Park   
30                Cedonia             1.0                  Park   
31   Central Park Heights             1.0                  Park   
49         East Arlington             1.0          Liquor Store   
73         Graceland Park             1.0            Sports Bar   
79              Gwynn Oak             1.0                  Park   
85            Harlem Park             1.0                  Park   
97               Idlewood             1.0                  Park   
119          Lucille Park             1.0                  Park   
154              Pen Lucy             1.0                   Bar   
182             St Helena             1.0                   Bar   
194           Violetville             1.0              Building   
204            West Hills             1.0                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                    Bar        Breakfast Spot  Fast Food Restaurant   
30                    Bar         Deli / Bodega           Yoga Studio   
31                    Spa                Market        Farmers Market   
49          Auto Workshop                  Park                  Farm   
73                    Bar           Yoga Studio  Fast Food Restaurant   
79       Business Service           Yoga Studio            Food Stand   
85                    Bar   American Restaurant              Bus Stop   
97            Yoga Studio     Convenience Store         Event Service   
119          Liquor Store         Deli / Bodega   Arts & Crafts Store   
154           Yoga Studio            Food Truck         Event Service   
182       Border Crossing           Social Club                  Port   
194           Music Venue                  Park           Yoga Studio   
204           Yoga Studio     Convenience Store         Event Service   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19            Event Space               Exhibit    Falafel Restaurant   
30   Fast Food Restaurant           Event Space               Exhibit   
31          Event Service           Event Space               Exhibit   
49   Ethiopian Restaurant         Event Service           Event Space   
73            Event Space               Exhibit    Falafel Restaurant   
79          Event Service           Event Space               Exhibit   
85             Playground        Farmers Market         Event Service   
97            Event Space               Exhibit    Falafel Restaurant   
119           Pizza Place                Market           Flower Shop   
154           Event Space               Exhibit    Falafel Restaurant   
182                  Park  Fast Food Restaurant                  Farm   
194         Event Service           Event Space               Exhibit   
204           Event Space               Exhibit    Falafel Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19                   Farm        Farmers Market            Yoga Studio  
30     Falafel Restaurant                  Farm         Farmers Market  
31     Falafel Restaurant                  Farm            Yoga Studio  
49                Exhibit    Falafel Restaurant         Farmers Market  
73                   Farm        Farmers Market                  Field  
79     Falafel Restaurant                  Farm         Farmers Market  
85            Event Space               Exhibit     Falafel Restaurant  
97                   Farm        Farmers Market   Fast Food Restaurant  
119           Flea Market           Fish Market                  Field  
154                  Farm        Farmers Market   Fast Food Restaurant  
182         Event Service           Event Space                Exhibit  
194    Falafel Restaurant                  Farm         Farmers Market  
204                  Farm        Farmers Market   Fast Food Restaurant

In [77]:
baltimore2_merged.loc[baltimore2_merged['Cluster Labels'] == 2, baltimore2_merged.columns[[0] + list(range(4, baltimore2_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
0          Abell             2.0           Pizza Place   
1      Allendale             2.0                   Bar   
2        Arcadia             2.0    Chinese Restaurant   
3      Arlington             2.0  Caribbean Restaurant   
6      Ashburton             2.0                  Lake   
..           ...             ...                   ...   
215     Woodmoor             2.0         Event Service   
216     Woodring             2.0        Baseball Field   
217   Wyman Park             2.0                Bakery   
218    Wyndhurst             2.0                  Bank   
220    York Road             2.0  Fast Food Restaurant   

       2nd Most Common Venue            3rd Most Common Venue  \
0         Chinese Restaurant                   Sandwich Place   
1        American Restaurant                 Basketball Court   
2          Other Repair Shop                      Yoga Studio   
3             Discount Store  Southern / Soul Food Restaurant   
6    Health & Beauty Service              Arts & Crafts Store   
..                       ...                              ...   
215         Business Service                      Yoga Studio   
216        German Restaurant                      Yoga Studio   
217              Pizza Place                        Pet Store   
218                      Spa                      Coffee Shop   
220   Thrift / Vintage Store                 Business Service   

      4th Most Common Venue   5th Most Common Venue 6th Most Common Venue  \
0                  Pharmacy                Bus Stop           Coffee Shop   
1               Yoga Studio                   Field               Exhibit   
2         Electronics Store           Event Service           Event Space   
3    Thrift / Vintage Store   Check Cashing Service    Chinese Restaurant   
6         Electronics Store           Event Service           Event Space   
..                      ...                     ...                   ...   
215       Convenience Store             Event Space               Exhibit   
216    Ethiopian Restaurant             Event Space               Exhibit   
217       Indian Restaurant                    Café           Coffee Shop   
218             Flower Shop  Furniture / Home Store      Toy / Game Store   
220     Fried Chicken Joint             Coffee Shop        Farmers Market   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                    Café         Grocery Store     Convenience Store   
1      Falafel Restaurant                  Farm        Farmers Market   
2                 Exhibit    Falafel Restaurant                  Farm   
3                Pharmacy     Convenience Store      Department Store   
6                 Exhibit    Falafel Restaurant                  Farm   
..                    ...                   ...                   ...   
215    Falafel Restaurant                  Farm        Farmers Market   
216    Falafel Restaurant                  Farm        Farmers Market   
217           Yoga Studio             Bookstore         Shopping Mall   
218              Pharmacy        Clothing Store         Grocery Store   
220         Event Service           Event Space               Exhibit   

    10th Most Common Venue  
0                Bookstore  
1     Fast Food Restaurant  
2           Farmers Market  
3            Design Studio  
6           Farmers Market  
..                     ...  
215   Fast Food Restaurant  
216   Fast Food Restaurant  
217             Shoe Store  
218          Women's Store  
220     Falafel Restaurant  

[176 rows x 12 columns]

In [78]:
baltimore2_merged.loc[baltimore2_merged['Cluster Labels'] == 3, baltimore2_merged.columns[[0] + list(range(4, baltimore2_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
23  Callaway-Garrison             3.0         Grocery Store   
88  Heritage Crossing             3.0         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23           Yoga Studio            Food Stand  Ethiopian Restaurant   
88           Yoga Studio            Food Stand  Ethiopian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23         Event Service           Event Space               Exhibit   
88         Event Service           Event Space               Exhibit   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23    Falafel Restaurant                  Farm         Farmers Market  
88    Falafel Restaurant                  Farm         Farmers Market

In [79]:
baltimore2_merged.loc[baltimore2_merged['Cluster Labels'] == 4, baltimore2_merged.columns[[0] + list(range(4, baltimore2_merged.shape[1]))]]

Neighborhood  Cluster Labels 1st Most Common Venue  \
87   Hawkins Point             4.0      Business Service   
146  Orchard Ridge             4.0          Home Service   
219   Yale Heights             4.0          Home Service   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
87           Home Service           Yoga Studio     Electronics Store   
146               Brewery           Yoga Studio         Event Service   
219           Yoga Studio   Dumpling Restaurant  Ethiopian Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
87          Event Service           Event Space               Exhibit   
146           Event Space               Exhibit    Falafel Restaurant   
219         Event Service           Event Space               Exhibit   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
87     Falafel Restaurant                  Farm         Farmers Market  
146                  Farm        Farmers Market   Fast Food Restaurant  
219    Falafel Restaurant                  Farm         Farmers Market